### DISCUSSION TITLE
PDM distribution = false + tool.pdm.scripts of type call

### DISCUSSION THREAD

**Discussion by jakeswenson** [pdm-project/pdm#3225] [Q&A]
I’m developing an application and want to use tool.pdm.scripts to call some of my app’s functions. Since this is an app and not a library, I’ve set:
```toml
[tool.pdm]
distribution = false
```

Given this setup, is it possible to have `call` scripts that reference modules in my code?

```toml
[tool.pdm.scripts]
extract.call = "foo.extract:main()"
```

When I run `pdm run extract`, I receive an error stating it can’t import `foo.extract`.

However, if I change distribution to true and then run `pdm install --dev`, everything works as expected. It seems my modules aren’t importable unless they’re installed into the virtual environment.

Is there a way to make this work without setting `distribution = true`? Should I consider opening an issue for this?

**Comment by TayfurYldz** [comment]
What you're seeing follows from `distribution = false`: PDM deliberately does not install the project itself when synchronizing that environment. A `call` script imports the target as a Python module; it doesn't implicitly add your source tree/package directory to Python's import path.

So `foo.extract:main()` works with `distribution = true` because the project gets installed (normally editable for development), making `foo` importable.

For an application that you intentionally don't want installed as a distribution, make the import path explicit for that script. For a conventional project where `foo/` is directly under the project root, for example:

```toml
[tool.pdm.scripts]
extract = { call = "foo.extract:main", env = { PYTHONPATH = "." } }
```

If you use a `src/` layout, use `PYTHONPATH = "src"` instead.

I wouldn't treat the current behavior itself as a bug: current PDM documentation explicitly says that a project configured with `distribution = false` is not installed itself. The choice is therefore either install the project and get normal package import semantics, or keep it non-distributable and explicitly make its source importable for the command.
